README.md
# Rampant Arsenal

Adds a variety of turrets to wage war. Adds Additional Ammo Types, Landmines, Equipment, Capsules, Infinite Technologies.

# Credits

YuokiTani - for turret sprite sheets

# Graphics License

THE GRAPHICS INCLUDED UNDER yuokiTani FOLDERS BELONG TO YuokiTani AND PERMISSION FOR THERE
USE MUST BE SOUGHT FROM YuokiTani.

# Forum Post

https://forums.factorio.com/viewtopic.php?f=190&t=60626

# Features
Advanced Laser Turret - Shoots an electric beam that pierces through enemies and does AoE damage.
Lightning Turret - Able to shoot beams at many biters at once.
Cannon Turret - Shoots cannon shells that that works with standard cannon ammo.
Rapid Cannon Turret - Shoots cannon shells at a high rate of fire.
Shotgun Turret - Shoots shotgun shells for cone AoE damage.
Capsule Turret - Shoots poison, slowdown, grenade, cluster grenades, landmines, distractors, defenders, and destroyers
Advanced Flamethrower - Shoots flame at many biters at the same time.
Rocket Turret - Shoots homing rocket projectiles.
Rapid Rocket Turret - Rapidly Shoots homing rocket projectiles.
Medic Turret - For healing the area surrounding the turret
Gun Turret MkII - For when gun turrets need more damage, range, and rate of fire
Rifle turret - For a tier 0 turret for before you have completed any research
Reactive Mending Walls & Gates - When they take damage they emit a repair cloud around themselves
Reinforced concrete walls and gates
Power Armor Mk 3
nuclear train
stronger car and a nuclear version
stronger tank and a nuclear version
Oil burner for generating electricity

High Explosives: landmines, grenades, bullets, rockets, shotgun shells, cannon shells, and artillery shells
Incendiary: landmines, grenades, bullets, rockets, shotgun shells, cannon shells, and artillery shells
Bio: landmines, grenades, bullets, rockets, shotgun shells, cannon shells, and artillery shells
Nuclear: landmines, and artillery shells

Mortar - for large range player combat using the capsule launcher ammo
Minigun - for an upgraded version of the smg with increased damage, range, and rate of fire
Health Capsule - Restore Health of the user
Speed Booster Capsule - for early game speed boosts
Upgraded Rocket Launcher - Increased damage, range, and rate of fire

Belt Immunity Equipment
Personal equipment tier 3 generator, battery, shield
Personal defense cannon, shotgun, lightning, bullets, and slowing equipment

Napalm - for an upgraded liquid to be used in fluid turrets

Adds Infinite Technology for the characters health
Adds Infinite Technology associated with each turret type and ammo type

### Four Technology Branches

High Explosives for AoE

Bio Weapons for Damage Over Time

Incendiary Weapons for combo over AoE and Damage over Time

Regeneration for weapons that provide healing, speed boosts, or mend themselves over time

# Version History

0.16.0 -
- Initial Release


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.1.7
  Tweaks:
    - Increased mending wall and gate detection range to 60
    - Increased research costs of high explosive, incendiary, and bio researches by ~4x
    - Increased research costs of new turrets by ~2x
    - Increased research costs of new vehicles by ~2x to ~4x
    - Increased power armor mk3 build time to 90 seconds and ~2x steel, processing chip cost, and added electric engines
    - Reduced gun turret MKII damage bonus by 60%

---------------------------------------------------------------------------------------------------
Version: 1.1.6
  Compatibility:
    - Added bobs warfare pierce and plasma to mending and reinforced walls and gates
  Bugfixes:
    - Mod setting for disabling vehicles now works with Krastorio

---------------------------------------------------------------------------------------------------
Version: 1.1.5
  Bugfixes:
    - Fixed mushroom cloud and fixed better artillery sounds compatibility

---------------------------------------------------------------------------------------------------
Version: 1.1.4
Date: 18. 02. 2022
  Contribution:
    - Locale for KO from x2605
  Tweaks:
    - Reduced advanced laser energy consumption by 55%
    - Increased advanced laser cooldown by 200%
    - Increased nuclear train consumption by 1000%
    - Increased advanced tank health by 1000 HP
    - Increased nuclear tank health by 2500 HP
    - Decreased nuclear tank consumption by 50%
    - Decreased advanced car consumption by 33%
    - Decreased nuclear car consumption by 33%
    - Increased nuclear train top speed by 10%
    - Changed license to GPLv3
  Bugfixes:
    - Fixed toxic launcher ammo not doing damage
    - Fixed Krastorio 2 overriding nuclear car and tank fuel type

---------------------------------------------------------------------------------------------------
Version: 1.1.3
Date: 16. 01. 2022
  Bugfixes:
    - Fixed infinite technologies requiring 1 science pack for first level of infinite tech

---------------------------------------------------------------------------------------------------
Version: 1.1.2
Date: 28. 12. 2021
  Bugfixes:
    - Corrected chinese locale duplicate key artillery remote

---------------------------------------------------------------------------------------------------
Version: 1.1.1
Date: 05. 09. 2021
  Contributions:
    - PlexPt has provided a Chinese locale
  Tweaks:
    - Increased HE artillery shell damage by ~15%
    - Changed behavior for incendiary and bio artillery shells to allow for damage scaling effect from research.
  Bugfixes:
    - Incendiary and Bio artillery shells not scaling with damage research
    - Fixed naplam not showing fire stats in tooltip window

---------------------------------------------------------------------------------------------------
Version: 1.1.0
Date: 05. 09. 2021
  Contributions:
    - moolex fixed unchecking use equipment breaking on lite artillery technology
  Tweaks:
    - Nuclear landmines no longer hurt the same force that laid them (Thanks Necrobotic)
    - Increased advCar consumption by 3x, lowered top speed, and set emissions to 15 minute
    - Increased advTank consumption by 2x, lowered top speed, and set emissions to 20 minute
    - Increased nuclearTank consumption by 5x, lowered top speed, and set emissions to 30 minute
    - Increased nuclearCar consumption by 3x, lowered top speed, and set emissions to 20 minute
    - Reduced mortar cooldown by %50
    - Increased the rate that mending wall events are processed to every 5 ticks

---------------------------------------------------------------------------------------------------
Version: 1.0.9
Date: 24. 04. 2021
  Integrations:
    - When RPGSystem is active the health research from Rampant arsenal is hidden
  Tweaks:
    - Increased rocket turrets minimum ranges by 5 and max range by 2
    - Reduced shotgun turret damage bonus by 50%
    - Increased min range on capsule turret by 5 and max range by 3
  Bugfixes:
    - Added check for artillery sound changes

---------------------------------------------------------------------------------------------------
Version: 1.0.8
Date: 17. 04. 2021
  Contributions:
    - MakaronP has provided an bugfix for the updated RU locale

---------------------------------------------------------------------------------------------------
Version: 1.0.7
Date: 16. 04. 2021
  Contributions:
    - Astorin has provided an updated RU locale

---------------------------------------------------------------------------------------------------
Version: 1.0.6
Date: 10. 04. 2021
  Improvements:
    - Added light artillery that shoots launcher ammo that is modeled after an artillery turret with remote
  Migration:
    - Fuel emission multipliers moved to Rampant Industry
  Tweaks:
    - Shotgun turret range reduced by 29%
    - Reverted nuclear cell to vanilla stack size
    - Capsule turret damage bonus reduced by 50%
    - Increased breaking power for advanced and nuclear car and tank
    - Reduced nuclear tank speed by x4
    - Increased nuclear tank fuel consumption by x3
    - Increased nuclear car fuel consumption by x6.6
    - Increased nuclear car fuel consumption by x3.33

---------------------------------------------------------------------------------------------------
Version: 1.0.5
Date: 28. 03. 2021
  Bugfixes:
    - Fixed nuclear generator not working in spidertron

---------------------------------------------------------------------------------------------------
Version: 1.0.4
Date: 28. 02. 2021
  Bugfixes:
    - Fixed air filter entity description
    - Fixed speed capsule effect description

---------------------------------------------------------------------------------------------------
Version: 1.0.3
Date: 07. 02. 2021
  Bugfixes:
    - Added increased healing capsule to 300 health recovery

prototypes/Medic.lua
-- Copyright (C) 2022  veden

-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.

-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.

-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <https://www.gnu.org/licenses/>.


local medic = {}

local sounds = require("__base__.prototypes.entity.sounds")
local recipeUtils = require("utils/RecipeUtils")
local technologyUtils = require("utils/TechnologyUtils")
local projectileUtils = require("utils/ProjectileUtils")
local ammoUtils = require("utils/AmmoUtils")
local turretUtils = require("utils/TurretUtils")

local makeAmmoTurret = turretUtils.makeAmmoTurret
local makeBombWave = projectileUtils.makeBombWave
local makeAmmo = ammoUtils.makeAmmo
local addEffectToTech = technologyUtils.addEffectToTech
local makeRecipe = recipeUtils.makeRecipe

local function medicSheet()
    return
        {
            layers =
                {
                    {
                        filename = "__RampantArsenal__/graphics/yuokiTani/entities/zone-expander.png",
                        priority = "high",
                        width = 96,
                        height = 128,
                        line_length = 4,
                        scale = 1,
                        run_mode = "forward",
                        direction_count = 1,
                        frame_count = 16,
                        shift = {-0.2, -1},
                    }
                }
        }
end


function medic.enable()

    data:extend(
        {
            {
                type = "ammo-category",
                name = "turret-capsule",
            }
    })

    local medicTurretAttributes = {
        name = "medic",
        icon = "__RampantArsenal__/graphics/icons/zone-expander.png",
        miningTime = 1,
        health = 1000,
        order = "b[turret]-a[zzzzmedic-turret]",
        collisionBox = {{-1, -1 }, {1, 1}},
        selectionBox = {{-1.2, -1.2 }, {1, 1}},
        hasBaseDirection = true,
        foldedAnimation = medicSheet(),
        foldingAnimation = medicSheet(),
        preparedAnimation = medicSheet(),
        preparingAnimation = medicSheet(),
        resistances = {
            {
                type = "fire",
                percent = 60
            },
            {
                type = "impact",
                percent = 30
            },
            {
                type = "explosion",
                percent = 30
            },
            {
                type = "physical",
                percent = 30
            },
            {
                type = "acid",
                percent = 30
            },
            {
                type = "electric",
                percent = 60
            },
            {
                type = "laser",
                percent = 60
            },
            {
                type = "poison",
                percent = 30
            }
        }
    }
    local medicTurret,medicTurretItem = makeAmmoTurret(medicTurretAttributes,
                                                       {
                                                           type = "projectile",
                                                           ammo_category = "turret-capsule",
                                                           cooldown = 1340,
                                                           warmup = 100,
                                                           projectile_creation_distance = 1,
                                                           damage_modifier = 1,
                                                           projectile_center = {0, 0},
                                                           range = 28,
                                                           sound = sounds.light_gunshot,
    })

    makeRecipe({
            name = medicTurretItem,
            icon = "__RampantArsenal__/graphics/icons/zone-expander.png",
            enabled = false,
            category = "crafting",
            ingredients = {
                {"engine-unit", 5},
                {"advanced-circuit", 5},
                {"steel-plate", 15}
            },
            result = medicTurretItem
    })


    local repairCapsules = makeAmmo({
            name = "self-repair-capsule",
            icon = "__RampantArsenal__/graphics/icons/medic-repair-pack.png",
            subgroup = "launcher-capsule",
            order = "y[turret-capsule]",
            magSize = 25,
            stackSize = 200,
            ammoType = {
                category = "turret-capsule",
                target_type = "position",
                clamp_position = true,

                action =
                    {
                        type = "direct",
                        action_delivery =
                            {
                                type = "projectile",
                                max_range = 0,
                                starting_speed = 1,
                                projectile = makeBombWave(
                                    {
                                        name = "repair"
                                    },
                                    {
                                        type = "direct",
                                        action_delivery =
                                            {
                                                type = "instant",
                                                target_effects =
                                                    {
                                                        {
                                                            type = "create-entity",
                                                            entity_name = "massive-repair-cloud-rampant-arsenal",
                                                            show_in_tooltip = true
                                                        }
                                                    }
                                            }
                                })
                            }
                    }
    }})

    makeRecipe({
            name = repairCapsules,
            icon = "__RampantArsenal__/graphics/icons/medic-repair-pack.png",
            enabled = false,
            ingredients = {
                {"copper-plate", 2},
                {"repair-capsule-rampant-arsenal", 5},
                {"advanced-circuit", 2}
            },
            result = repairCapsules
    })

    addEffectToTech("regeneration-turrets",
                    {
                        type = "unlock-recipe",
                        recipe = medicTurretItem
    })

    addEffectToTech("regeneration-turrets",
                    {
                        type = "unlock-recipe",
                        recipe = repairCapsules
    })


    addEffectToTech((settings.startup["rampant-arsenal-hideVanillaDamageTechnologies"].value and "capsule-turret-damage-1") or "stronger-explosives-1",
                    {
                        type = "turret-attack",
                        turret_id = medicTurret,
                        modifier = 0.1
    })

    addEffectToTech((settings.startup["rampant-arsenal-hideVanillaDamageTechnologies"].value and "capsule-turret-damage-2") or "stronger-explosives-2",
                    {
                        type = "turret-attack",
                        turret_id = medicTurret,
                        modifier = 0.1
    })

    addEffectToTech((settings.startup["rampant-arsenal-hideVanillaDamageTechnologies"].value and "capsule-turret-damage-3") or "stronger-explosives-3",
                    {
                        type = "turret-attack",
                        turret_id = medicTurret,
                        modifier = 0.2
    })

    addEffectToTech((settings.startup["rampant-arsenal-hideVanillaDamageTechnologies"].value and "capsule-turret-damage-4") or "stronger-explosives-4",
                    {
                        type = "turret-attack",
                        turret_id = medicTurret,
                        modifier = 0.3
    })

    addEffectToTech((settings.startup["rampant-arsenal-hideVanillaDamageTechnologies"].value and "capsule-turret-damage-5") or "stronger-explosives-5",
                    {
                        type = "turret-attack",
                        turret_id = medicTurret,
                        modifier = 0.3
    })

    addEffectToTech((settings.startup["rampant-arsenal-hideVanillaDamageTechnologies"].value and "capsule-turret-damage-6") or "stronger-explosives-6",
                    {
                        type = "turret-attack",
                        turret_id = medicTurret,
                        modifier = 0.4
    })

    if (settings.startup["rampant-arsenal-useInfiniteTechnologies"].value) then
        addEffectToTech((settings.startup["rampant-arsenal-hideVanillaDamageTechnologies"].value and "capsule-turret-damage-7") or "stronger-explosives-7",
                        {
                            type = "turret-attack",
                            turret_id = medicTurret,
                            modifier = 0.2
        })
    end

    addEffectToTech("stronger-explosives-1",
                    {
                        type = "ammo-damage",
                        ammo_category = "turret-capsule",
                        modifier = 0.2
    })

    addEffectToTech("stronger-explosives-2",
                    {
                        type = "ammo-damage",
                        ammo_category = "turret-capsule",
                        modifier = 0.2
    })

    addEffectToTech("stronger-explosives-3",
                    {
                        type = "ammo-damage",
                        ammo_category = "turret-capsule",
                        modifier = 0.3
    })

    addEffectToTech("stronger-explosives-4",
                    {
                        type = "ammo-damage",
                        ammo_category = "turret-capsule",
                        modifier = 0.3
    })

    addEffectToTech("stronger-explosives-5",
                    {
                        type = "ammo-damage",
                        ammo_category = "turret-capsule",
                        modifier = 0.4
    })

    addEffectToTech("stronger-explosives-6",
                    {
                        type = "ammo-damage",
                        ammo_category = "turret-capsule",
                        modifier = 0.2
    })

    if (settings.startup["rampant-arsenal-useInfiniteTechnologies"].value) then
        addEffectToTech("stronger-explosives-7",
                        {
                            type = "ammo-damage",
                            ammo_category = "turret-capsule",
                            modifier = 0.2
        })
    end

end

return medic
